vmsyscall/src/memory_map.rs

```diff
@@ -0,0 +1,317 @@
+        self.iter_mut().for_each(|r| {
+            if r.region_type == region_type {
+                r.region_type = MemoryRegionType::Usable
+            }
+        });
+    }
+
+    /// Add a region to the `MemoryMap`.
+    pub fn add_region(&mut self, region: MemoryRegion) {
+        if let Err(()) = self.entries.iter_mut().try_for_each(|last_region| {
+            if last_region.region_type == region.region_type
+                && last_region.range.end_frame_number >= region.range.start_frame_number
+                && last_region.range.end_frame_number <= region.range.end_frame_number
+            {
+                last_region.range.end_frame_number = region.range.end_frame_number;
+                return Err(());
+            }
+            Ok(())
+        }) {
+            return;
+        };
+
+        assert!(
+            self.next_entry_index() < MAX_MEMORY_MAP_SIZE,
+            "too many memory regions in memory map"
+        );
+
+        self.entries[self.next_entry_index()] = region;
+        self.next_entry_index += 1;
+        self.sort();
+    }
+
+    /// Mark a region as allocated.
+    pub fn mark_allocated_region(&mut self, region: MemoryRegion) {
+        let mut region = region;
+        for r in self.iter_mut() {
+            // New region inside region of same type
+            if r.region_type == region.region_type
+                && r.range.start_frame_number <= region.range.start_frame_number
+                && r.range.end_frame_number >= region.range.end_frame_number
+            {
+                return;
+            }
+
+            // New region extends old region
+            if r.region_type == region.region_type
+                && r.range.start_frame_number <= region.range.start_frame_number
+                && r.range.end_frame_number > region.range.start_frame_number
+                && r.range.end_frame_number <= region.range.end_frame_number
+            {
+                region.range.start_frame_number = r.range.end_frame_number;
+            }
+
+            if region.range.start_frame_number >= r.range.end_frame_number {
+                continue;
+            }
+            if region.range.end_frame_number <= r.range.start_frame_number {
+                continue;
+            }
+
+            if r.region_type != MemoryRegionType::Usable {
+                panic!(
+                    "region {:x?} overlaps with non-usable region {:x?}",
+                    region, r
+                );
+            }
+
+            match region
+                .range
+                .start_frame_number
+                .cmp(&r.range.start_frame_number)
+            {
+                Ordering::Equal => {
+                    if region.range.end_frame_number < r.range.end_frame_number {
+                        // Case: (r = `r`, R = `region`)
+                        // ----rrrrrrrrrrr----
+                        // ----RRRR-----------
```

Comment:
These comments are as good as ciphertext to me. I don't understand them.

Reply:
These are both regions, which start at the same point.. one is bigger than the other.